### DISCUSSION TITLE
New feature: add 'allow_override_connect' startup option

### DISCUSSION THREAD

**Discussion by RJiazhen** [peers/peerjs#1234] [Ideas & Feature requests]
Currently, when the peer server receives a connection request with the same connection ID as an existing one, it immediately returns an "ID-TAKEN" error and disconnects the connection. 

However, I would like to add an "allow_override_connect" startup option (defaulting to false). When set to true, **it should accept the connection request from the new client and send an "OVERRIDE_CONNECTION" error to the old client, then disconnect the old connection**. Like this:

I have forked the peer-server and implemented the related functionality. However, due to the limitations in PeerJS regarding error types, the client cannot respond to this error directly. As a temporary solution, I am overriding the _handleMessage method to handle this response.

So, can we consider adding relevant startup options in peer-server and implementing corresponding response functionality in PeerJS?

I would also be happy to participate in the development of this feature.

**Comment by jonasgloning** [comment]
Can you tell me more about your use case and your setup?

In the default case, the IDs of the connections are public and not authenticated, so you could take over  anyone's connections.

I'm very open to overhauling the whole ID/permission thing in the near future, allowing server-generated and properly authenticated IDs, and then give more freedom on what you can do in case of a conflict.

**Reply by RJiazhen** [reply]
I implemented a remote control function with peerjs. 

The remote control can control the player, and when the two terminals connect with peerjs, they will use fixed rules to generate id. 
For example, if both the remote control and the player belong to a session called `aaa`, they can connect using the two ID of `aaa_remote_control` and `aaa_player`, respectively. 

Normally, the communication between the two client is normal, but when the player is running on the new device (and the old player is not properly disconnected), the player on the new device cannot connect to the peer-server, and the remote control and the old player will not know that a new player has been created and let the remote control connect to the new player. 

The following is a picture of an example. 
![Frame 67](https://github.com/peers/peerjs/assets/103310338/eeb1a6c9-19e2-4c8b-b1f5-51f5cacb7409)

I know I can create another ws connection and pass this information through this ws connection, but that would be superfluous (peer-server has already established a ws connection).

**Comment by fix** [comment]
i think you can't escape the implementation of proper "hey this id is mine, here is the proof" ie public key cryptography even ephemeral for a session so you can claim "back" the id supposed to be already taken.
